### PR TITLE
fix: add Set-Cookie header for sam local

### PR DESCRIPTION
**Pull request by xazhao**
#### Which issue(s) does this change fix?
https://github.com/aws/aws-sam-cli/issues/3168

#### How does it address the issue?
Copy cookies from `cookies` field to `Set-Cookie` header

#### Checklist

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/local/apigw/local_apigw_service.py [hunk 1]
```diff
@@ -526,6 +527,20 @@ def _parse_v2_payload_format_lambda_output(lambda_output: str, binary_types, fla
...
+        if isinstance(json_output.get("cookies"), list):
+            # add all cookies to Set-Cookie header
+            if "Set-Cookie" not in headers:
+                headers["Set-Cookie"] = ""
+            for cookie in json_output.get("cookies"):
+                headers["Set-Cookie"] = headers["Set-Cookie"] + ";" + cookie
+            # remove the first semi colon and trim the header
+            if headers["Set-Cookie"][0] == ";":
+                headers["Set-Cookie"] = headers["Set-Cookie"][1:]
+            headers["Set-Cookie"] = headers["Set-Cookie"].strip()
```
Comment (aahung) [inline samcli/local/apigw/local_apigw_service.py:542]: nit: There are several branches that we can use "join" to simplify:

```suggestion
            headers["Set-Cookie"] = (
                ";".join([
                    cookie for cookie in [headers.get("Set-Cookie")] + json_output.get("cookies") if cookie
                ])
            ).strip()
```

File: tests/unit/local/apigw/test_local_apigw_service.py [hunk 2]
```diff
@@ -1112,6 +1112,30 @@ def test_custom_content_type_header_is_not_modified(self):
         self.assertIn("Content-Type", headers)
         self.assertEqual(headers["Content-Type"], "text/xml")
 
+    def test_custom_cookies_in_payload_format_version_2(self):
+        lambda_output = (
+            '{"statusCode": 200, "cookies": ["cookie1=test1", "cookie2=test2"], "body": "{\\"message\\":\\"Hello from Lambda\\"}", '
+            '"isBase64Encoded": false}'
+        )
```
Comment (aahung) [inline tests/unit/local/apigw/test_local_apigw_service.py:1119]: can we add a test if "Set-Cookie" pre-exists in the headers?

Comment (xazhao) [reply tests/unit/local/apigw/test_local_apigw_service.py:1119]: Sure! Added that test

File: samcli/local/apigw/local_apigw_service.py [hunk 3]
```diff
@@ -506,6 +506,7 @@ def _parse_v2_payload_format_lambda_output(lambda_output: str, binary_types, fla
         :return: Tuple(int, dict, str, bool)
         """
         # pylint: disable-msg=too-many-statements
+        # pylint: disable=too-many-branches
```
Comment (aahung) [inline samcli/local/apigw/local_apigw_service.py:509]: Do we still need this?

**Review by aahung — APPROVED** [review]
LGTM

File: samcli/local/apigw/local_apigw_service.py (hunk 3 above)
Comment (xazhao) [reply samcli/local/apigw/local_apigw_service.py]: Nope. Deleted this line

File: samcli/local/apigw/local_apigw_service.py [hunk 4]
```diff
@@ -526,6 +526,15 @@ def _parse_v2_payload_format_lambda_output(lambda_output: str, binary_types, fla
...
         headers = Headers(json_output.get("headers") or {})
 
+        # cookies is a new field in payload format version 2.0 (a list)
+        # https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html
+        # we need to move cookies to Set-Cookie headers.
+        if isinstance(json_output.get("cookies"), list):
+            # add all cookies to Set-Cookie header
+            headers["Set-Cookie"] = (
+                ";".join([cookie for cookie in [headers.get("Set-Cookie")] + json_output.get("cookies") if cookie])
+            ).strip()
```
Comment (hoffa) [inline samcli/local/apigw/local_apigw_service.py:536]: Please add a reference as to why this `Set-Cookie` format is correct.

In [RFC 6265](https://datatracker.ietf.org/doc/html/rfc6265#section-4.1.1) the syntax doesn't describe allowing concatenating cookies with `;`. Instead, cookies are on separate `Set-Cookie` values.

From [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie):

> To send multiple cookies, multiple `Set-Cookie` headers should be sent in the same response.

Comment (xazhao) [reply samcli/local/apigw/local_apigw_service.py]: Yep you're right. API Gateway only supported single `Set-Cookie` header some time ago and I read a outdated document. So I was trying to make it look like what API Gateway should do. I just found API Gateway can support multiple `Set-Cookie` headers now. Updated the code.

File: samcli/local/apigw/local_apigw_service.py [hunk 5]
```diff
@@ -526,6 +526,15 @@ def _parse_v2_payload_format_lambda_output(lambda_output: str, binary_types, fla
         status_code = json_output.get("statusCode") or 200
         headers = Headers(json_output.get("headers") or {})
 
+        # cookies is a new field in payload format version 2.0 (a list)
+        # https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html
+        # we need to move cookies to Set-Cookie headers.
+        if isinstance(json_output.get("cookies"), list):
```
[the remark this answers is not in the document]
Comment (xazhao) [reply samcli/local/apigw/local_apigw_service.py]: Yep updated the code.

**Review by hoffa — APPROVED** [review]

**Pull request merged**